client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,343 @@
+import { getVariant } from "../data/gameData";
+import * as deckRules from "../rules/deck";
+import { GameAction } from "../types/actions";
+import CardState from "../types/CardState";
+import GameMetadata from "../types/GameMetadata";
+import Variant from "../types/Variant";
+
+export default function cardDeductionReducer(
+  deck: readonly CardState[],
+  oldDeck: readonly CardState[],
+  action: GameAction,
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+): CardState[] {
+  switch (action.type) {
+    case "cardIdentity":
+    case "clue":
+    case "discard":
+    case "play":
+    case "draw": {
+      return makeDeductions(deck, oldDeck, hands, metadata);
+    }
+    default: {
+      return Array.from(deck);
+    }
+  }
+}
+
+function makeDeductions(
+  deck: readonly CardState[],
+  oldDeck: readonly CardState[],
+  hands: ReadonlyArray<readonly number[]>,
+  metadata: GameMetadata,
+) {
+  const newDeck: CardState[] = Array.from(deck);
+  const variant = getVariant(metadata.options.variantName);
+  const cardCountMap = getCardCountMap(variant);
+
+  // We need to calculate our own unknown cards first because those possibilities will be needed for
+  // pretending like we know what the other players see.
+  calculatePlayerPossibilities(
+    metadata.ourPlayerIndex,
+    metadata.ourPlayerIndex,
+    hands,
+    newDeck,
+    oldDeck,
+    cardCountMap,
+  );
+  for (let playerIndex = 0; playerIndex < hands.length; playerIndex++) {
+    if (playerIndex !== metadata.ourPlayerIndex) {
+      calculatePlayerPossibilities(
+        playerIndex,
+        metadata.ourPlayerIndex,
+        hands,
+        newDeck,
+        oldDeck,
+        cardCountMap,
+      );
+    }
+  }
+  return newDeck;
+}
+
+function calculatePlayerPossibilities(
+  playerIndex: number,
+  ourPlayerIndex: number,
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  oldDeck: readonly CardState[],
+  cardCountMap: readonly number[][],
+) {
+  const cardsToCalculate = getCardsToCalculate(
```

Comment:
I got rid of the cardsToCalculate and unknownCards list in favor of iterating directly.  The generateDeckPossibilities method (previously generatePossibilitiesForUnknownCards) still creates a list for deckPossibilities, since that's needed to sort the values within before iteration.